transports/quic/src/transport.rs

```diff
@@ -191,18 +214,65 @@ impl<P: Provider> Transport for GenTransport<P> {
                 &mut listeners[index].dialer_state
             }
         };
-        Ok(dialer_state.new_dial(socket_addr, self.handshake_timeout, version))
+        Ok(dialer_state.new_dial(socket_addr, handshake_timeout, version))
     }
 
     fn dial_as_listener(
         &mut self,
         addr: Multiaddr,
     ) -> Result<Self::Dial, TransportError<Self::Error>> {
-        // TODO: As the listener of a QUIC hole punch, we need to send a random UDP packet to the
-        // `addr`. See DCUtR specification below.
-        //
-        // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
-        Err(TransportError::MultiaddrNotSupported(addr))
+        let (socket_addr, _version, peer_id) = self.remote_multiaddr_to_socketaddr(addr.clone())?;
+
+        let peer_id = peer_id.ok_or(TransportError::MultiaddrNotSupported(addr))?;
+
+        let listeners = self.eligible_listeners(&socket_addr);
+
+        let endpoint_channel = match listeners.len() {
+            0 => {
+                return Err(TransportError::Other(
+                    Error::NoActiveListenerForDialAsListener,
+                ));
+            }
+            1 => listeners[0].endpoint_channel.clone(),
+            _ => {
+                // Pick any listener to use for dialing.
+                // We hash the socket address to achieve determinism.
+                let mut hasher = DefaultHasher::new();
+                socket_addr.hash(&mut hasher);
+                let index = hasher.finish() as usize % listeners.len();
+                listeners[index].endpoint_channel.clone()
+            }
+        };
+
+        let hole_puncher = HolePuncher::new(endpoint_channel, socket_addr, self.handshake_timeout);
```

Comment:
If I understand correctly `HolePuncher` will operate untile `self.handshake_timeout`. We might have already established the connection in the meantime.

What do you think of creating another `oneshot::channel`, passing the `receiver` into `HolePuncher` and the `sender` into the below returned `Box::pin(async{ ...})`. When the latter is dropped, i.e. the dial is done, the `HolePuncher` will receive a signal through the `receiver` that the `sender` side is dropped, thus knowing that it can terminate.

Reply:
The `HolePuncher` should get dropped once we receive the established connection details, right? At least with the `select!` that we have right now.

Reply:
Oh silly me. You are right. Thanks @arpankapoor.